pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java

```diff
@@ -49,57 +47,67 @@
-    _dataSchema = dataSchema;
-    _mailboxService = mailboxService;
+    this(mailboxService, dataTableBlockBaseOperator, receivingStageInstances, exchangeType, keySelector,
+        isLeafStageSender, server -> toMailboxId(server, jobId, stageId, hostName, port), BlockExchange::getExchange);
+  }
+
+  @VisibleForTesting
+  MailboxSendOperator(MailboxService<TransferableBlock> mailboxService,
+      Operator<TransferableBlock> dataTableBlockBaseOperator, List<ServerInstance> receivingStageInstances,
+      RelDistribution.Type exchangeType, KeySelector<Object[], Object[]> keySelector,
+      boolean isLeafStageSender, MailboxIdGenerator mailboxIdGenerator, BlockExchangeFactory blockExchangeFactory) {
     _dataTableBlockBaseOperator = dataTableBlockBaseOperator;
-    _exchangeType = exchangeType;
-    if (_exchangeType == RelDistribution.Type.SINGLETON) {
+
+    List<MailboxIdentifier> receivingMailboxes;
+    if (exchangeType == RelDistribution.Type.SINGLETON) {
+      // TODO: this logic should be moved into SingletonExchange
       ServerInstance singletonInstance = null;
       for (ServerInstance serverInstance : receivingStageInstances) {
-        if (serverInstance.getHostname().equals(_mailboxService.getHostname())
-            && serverInstance.getQueryMailboxPort() == _mailboxService.getMailboxPort()) {
+        if (serverInstance.getHostname().equals(mailboxService.getHostname())
+            && serverInstance.getQueryMailboxPort() == mailboxService.getMailboxPort()) {
           Preconditions.checkState(singletonInstance == null, "multiple instance found for singleton exchange type!");
           singletonInstance = serverInstance;
         }
       }
       Preconditions.checkNotNull(singletonInstance, "Unable to find receiving instance for singleton exchange");
-      _receivingStageInstances = Collections.singletonList(singletonInstance);
+      receivingMailboxes = Collections.singletonList(mailboxIdGenerator.generate(singletonInstance));
     } else {
-      _receivingStageInstances = receivingStageInstances;
+      receivingMailboxes = receivingStageInstances
+          .stream()
+          .map(mailboxIdGenerator::generate)
+          .collect(Collectors.toList());
     }
-    _keySelector = keySelector;
-    _serverHostName = hostName;
-    _serverPort = port;
-    _jobId = jobId;
-    _stageId = stageId;
-    Preconditions.checkState(SUPPORTED_EXCHANGE_TYPE.contains(_exchangeType),
-        String.format("Exchange type '%s' is not supported yet", _exchangeType));
-    _isLeafStageSender = isLeafStageSender;
+
+    BlockSplitter splitter = (block, type, size)
+        -> TransferableBlockUtils.splitBlock(block, type, size, isLeafStageSender);
+    _exchange = blockExchangeFactory.build(mailboxService, receivingMailboxes, exchangeType, keySelector, splitter);
+
+    Preconditions.checkState(SUPPORTED_EXCHANGE_TYPE.contains(exchangeType),
+        String.format("Exchange type '%s' is not supported yet", exchangeType));
```

Comment:
why are we checking here instead of at the beginning?

Reply:
no good reason, this is just what it was before my refactor. I'll change that.